README
This module gives you lots of flexibility when testing deep structres.


CHANGES
Revision history for Test-Deep

0.111     2013-11-30

        - When printing diagnostics, differentiate the type of a blessed object
          from the name of the class itself (RT#78288, caused by changes to how
          blessed objects are treated in 0.109) (thanks, Karen Etheridge)

        - Typo fixes (thanks, David Steinbrunner)

        - Fixes to clarity and accuracy of documentation (thanks, Michael Hamlin)

0.110     2012-06-16

          Allow methods() and listmethods() to work again on class methods
          (RT#77804) (thanks, Ricardo Signes!)

0.109     2012-02-16

          UNIVERSAL::isa and UNIVERAL::can are no longer called as functions:
          this was deprecated in UNIVERSAL 1.03. (RT#74760, Karen Etheridge)

          the code that builds the exporter configuration is cleaned up; more
          documentation of how it may change (for the better, without breaking
          stuff) should appear in the near future

0.108     2010-10-15

          simple (stringwise) comparison should now be much faster (thanks,
          NCLEATON)

0.107     2009-10-28

          New maintainer:  Ricardo Signes (rjbs)

          On 5.10.0 and newer perl, install to the "site" directory, not the
          core perl directory in @INC.  (Prior to 5.10.0, @INC ordering was
          unfortunate, and fixing the install target would make upgrading
          difficult.)

          cope with new stringification of qr{} objects in 5.13.x+ (thanks,
          ANDK)

          fix Pod escaping errors (thanks, FWIE)

0.106

          Release after dev.

0.105-dev

          Fix for perl 5.010.1 - the code to check the version and act
          correctly on regexp refs treated .1 as behaving like 5.011.

0.104

          Document behaviour of cmp_bag when a non ARRAY-ref argument is passed
          to it. Explicity test for this a die with a useful message.

          Document and export cmp_details and deep_diag, thanks to Tom Hukins
          for the patch.

0.103

          Detect whether isa() is being called with 1 or 2 arguments and
          dispatch to the correct function. This is hacky but fixes the problem
          of clashing with UNIVERSAL::isa().

0.102

          Behave well when a code comparator doesn't give diagnostics. Thanks
          to Alex Kapranoff.

0.101

          Fix an overloading problem with All and Any. Thanks to Kostas
          Chatzikokolakis for reporting the bug and especially for explaining
          how to fix it :)

          Added a nasty hack to smooth over a problem in Test::Tester.

0.100

          Apply patch from Andreas Koenig (ANDK) to cope with Perl 5.11's new
          REGEXP objects.

0.099

          Don't explode on perl's that don't have weakrefs. If they're not
          available, just don't weaken the refs. It means refs can be cached
          for longer than desirable but it's better than explodiing.

          This doesn't fix all of te problems with perl 5.005 but I'm pushing
          it out because the previous version number upsets CPAN.pm.

0.098

          Cleaned up Test::Deep::NoTest by making it call Test::Deep's import,
          that way it will always be in sync for exports. Also added some POD
          docs.

          Added notest_extra.t to check that things are correctly exported and
          working.

          Add cmp_details but left it undocumented. It manages localising the
          stack etc. Factored eq_deeply and cmp_deeply through that. This meant
          adding an icky hack into deep_diag. I should clean that whole thing
          up.

          Remove $name from params of eq_deeply, in code and docs, it was
          unused.

0.097

          Fixed doc typo for re().

          Fix diagnostics for code(), it was always printing undef instead of
          the got value.

0.096

          %WrapCache was keeping references to external data. It's now
          local()ised at the start of a comparison just like the other caches
          (why I didn't do that when I added it, I don't know). Thanks to
          Matthijs Bomhoff for reporting the problem.

0.095

          Docs and code didn't match, useclass was actually requireclass,
          available as both now

0.094

          Changed Set and Bag to no longer issue warnings when undefs are
          present. Needed to make the sort and the diagnotics
          undef-aware. Thanks to Colin Kuskie for pointing this out.

          Added tests for this.

0.093

          Fixed inifinte recursion when adding comparators into bags. That
          whole area is unpleasant - conceptually as well as
          implementation-wise. Comparators no longer inherit a compare method,
          so we only call the specialised compare if it exists, otherwise we
          just do normal deep comparison.

          Removed a debugging print.

0.092

          Got the if backwards for loading Test::Builder (again for NoTest -
          simples changes...). No ill effects for most people as they load it
          via something else any but ApacheSession for example doesn't.

          Now tetsing for lack of testing framework.

0.091

          Forgot to export lots of stuff via NoTest

0.090

          Added Test::Builder::NoTest so that eq_deeply can be use entirely
          without Test::Builder.

0.089

          Wasn't corrrectly setting $Test::Builder::Level in cmp_bag, cmp_set
          and cmp_methods, this caused test error messages to include the wrong
          line/file.

          Whitespace changes.

0.088

          Slight twiddle to silence warnings from the new Test::Tester thanks to
          SSORICHE

0.087

          cmp_methods, cmp_bag and cmp_set weren't passing on the test name.
          Thanks to Alex Kapranoff for the patch.

0.086

          Add $tolerance to Number so now you can check that 2 numbers are
          equal, give or take a little bit.

          Autogenerating most of the convenience constructor routines, deleted

Makefile.PL
use strict;

use ExtUtils::MakeMaker;

# this ugliness comes from incompatibility of certain versions of
# Test::Tester with certain version of Test::Builder. The problem is
# that people might have an old TT lying around and are also likely to
# have quite a new TB. This detects that situation and hopefully
# demands an install of a newer TT.

my $tt_prereq = "0.04";

if (eval { require Test::Tester; require Test::Builder; 1 } &&
  $Test::Tester::VERSION <= 0.106 &&
  $Test::Builder::VERSION >= 0.78) {

  $tt_prereq = "0.107";
}

my $mm_ver = ExtUtils::MakeMaker->VERSION;

WriteMakefile(
  AUTHOR       => 'Fergal Daly <fergal@esatclear.ie>',
  NAME         => 'Test::Deep',
  VERSION_FROM => './lib/Test/Deep.pm',
  PREREQ_PM    => {
    'Test::More'       => '0',
    'Test::Tester'     => $tt_prereq,
    'Test::NoWarnings' => '0.02',
    'Scalar::Util'     => '1.09',

    # apparently CPAN doesn't get the version of Scalar::Util
    'List::Util'       => '1.09',
  },

  # This is a dumb mistake.  Why did it get done?  I'm the one who did it and I
  # don't know.  It only affects perl 5.8, and stopping doing it now would be a
  # problem, because upgrades wouldn't actually upgrade.  I'll just leave it
  # here until 5.8 is really and truly dead enough. -- rjbs, 2013-11-30
  ($] < 5.010 ? (INSTALLDIRS => 'perl') : ()),

  ($mm_ver < 6.46 ? () : (META_MERGE => {
    resources => {
        license     =>      'http://dev.perl.org/licenses/',
        homepage    =>      'http://github.com/rjbs/Test-Deep/',
        bugtracker  =>      'http://github.com/rjbs/Test-Deep/issues/',
        repository  =>      'http://github.com/rjbs/Test-Deep/',
        MailingList =>      'http://lists.perl.org/list/perl-qa.html',
    },
  })),

);
